ambry-admin/src/main/java/com.github.ambry.admin/AdminBlobStorageService.java

```diff
@@ -572,63 +569,59 @@ public void onCompletion(final BlobInfo routerResult, Exception routerException)
+                  boolean blobNotModified = restResponseChannel.getStatus() == ResponseStatus.NotModified;
+                  try {
+                    if (securityException == null) {
+                      if (subResource != null) {
+                        BlobInfo blobInfo = routerResult.getBlobInfo();
+                        Map<String, String> userMetadata = RestUtils.buildUserMetadata(blobInfo.getUserMetadata());
+                        if (userMetadata == null) {
+                          restResponseChannel.setHeader(RestUtils.Headers.CONTENT_TYPE, "application/octet-stream");
+                          restResponseChannel
+                              .setHeader(RestUtils.Headers.CONTENT_LENGTH, blobInfo.getUserMetadata().length);
+                          response = new ByteBufferReadableStreamChannel(ByteBuffer.wrap(blobInfo.getUserMetadata()));
+                        } else {
+                          setUserMetadataHeaders(userMetadata, restResponseChannel);
+                          restResponseChannel.setHeader(RestUtils.Headers.CONTENT_LENGTH, 0);
+                          response = new ByteBufferReadableStreamChannel(AdminBlobStorageService.EMPTY_BUFFER);
+                        }
+                      } else if (!blobNotModified) {
+                        response = routerResult.getBlobDataChannel();
+                      } else {
+                        // If the blob was not modified, we need to close the channel, as it will not be submitted to
+                        // the RestResponseHandler
+                        routerResult.getBlobDataChannel().close();
+                      }
                     }
-                  } else if (!blobNotModified) {
-                    logger.trace("Forwarding GET after HEAD for {} to the router", blobId);
-                    router.getBlob(blobId, null, new GetCallback(restRequest, restResponseChannel));
+                  } catch (Exception e) {
+                    adminMetrics.getSecurityResponseCallbackProcessingError.inc();
+                    securityException = e;
+                  } finally {
+                    submitResponse(restRequest, restResponseChannel, response, securityException);
+                    securityCallbackTracker.markCallbackProcessingEnd();
                   }
                 }
-              } catch (Exception e) {
-                adminMetrics.getSecurityResponseCallbackProcessingError.inc();
-                securityException = e;
-              } finally {
-                if (response != null || securityException != null || blobNotModified) {
-                  submitResponse(restRequest, restResponseChannel, response, securityException);
-                }
-                securityCallbackTracker.markCallbackProcessingEnd();
-              }
-            }
-          });
+              });
         }
       } catch (Exception e) {
         adminMetrics.headForGetCallbackProcessingError.inc();
```

Comment:
It seems that I forgot to migrate some changes from `Ambry...` to `Admin...`, so I'll double check this.